zulipterminal/ui_tools/views.py

```diff
@@ -1215,34 +1216,38 @@ def __init__(self, controller: Any, msg: Message, title: str,
                     grouped_reactions[reaction] = user
             msg_info.append(('Reactions', list(grouped_reactions.items())))
 
-        popup_width, column_widths = self.calculate_table_widths(msg_info,
-                                                                 len(title))
-        widgets = self.make_table_with_categories(msg_info, column_widths)
+        self.popup_width, column_widths = (self.calculate_table_widths(
+                                           msg_info, len(title)))
+        self.widgets = self.make_table_with_categories(msg_info, column_widths)
 
         if message_links:
-            message_link_widgets = []
-            message_link_width = 0
-            for index, link in enumerate(message_links):
-                text, link_index, _ = message_links[link]
-                caption = '{}: {}\n{}'.format(link_index, text, link)
-                message_link_width = max(
-                    message_link_width,
-                    len(max(caption.split('\n'), key=len))
-                )
-
-                display_attr = None if index % 2 else 'popup_contrast'
-                message_link_widgets.append(
-                    MessageLinkButton(controller, caption, link, display_attr)
-                )
-
             # slice_index = Number of labels before message links + 1 newline
             #               + 1 'Message Links' category label.
-            slice_index = len(msg_info[0][1]) + 2
-            widgets = (widgets[:slice_index] + message_link_widgets
-                       + widgets[slice_index:])
-            popup_width = max(popup_width, message_link_width)
+            self.slice_index = len(msg_info[0][1]) + 2
+            self.set_links(message_links)
+
+        super().__init__(self.controller, self.widgets, 'MSG_INFO',
+                         self.popup_width, title)
+
+    def set_links(self,
+                  links: 'OrderedDict[str, Tuple[str, int, bool]]') -> None:
+        link_widgets = []
+        link_width = 0
+        for index, link in enumerate(links):
+            text, link_index, _ = links[link]
+            caption = '{}: {}\n{}'.format(link_index, text, link)
+            link_width = max(
+                link_width,
+                len(max(caption.split('\n'), key=len))
+            )
 
-        super().__init__(controller, widgets, 'MSG_INFO', popup_width, title)
+            display_attr = None if index % 2 else 'popup_contrast'
+            link_widgets.append(
+                MessageLinkButton(self.controller, caption, link, display_attr)
+            )
+        self.widgets = (self.widgets[:self.slice_index] + link_widgets
+                        + self.widgets[self.slice_index:])
+        self.popup_width = max(self.popup_width, link_width)
```

Comment:
You're doing a lot of setting of attributes here, but setting the values before calling - that suggests it may be clearer passing things in as parameters and/or back as return values. In this case can we leave these two lines in `__init__` and simplify this?